Reflectometry GUI restructuring 5: presenter communication

gemmaguest opened the pull request:
***PRs #23630, #24345, #24368 and #24374 should be merged first***

**Description of work.**
Ongoing work to integrate the `JobTreeView` into the `ISIS Reflectometry` GUI. This PR is being merged to a feature branch because not all functionality is implemented yet.

This PR adds cross-presenter communication:
- The Instrument combo boxes are now synchronised between the main `Runs` tab and its child `RunsTable` component.
- **To follow** Disable inputs across several different tabs when processing is in progress.

**Report to:** nobody

**To test:**

- Open the `ISIS Reflectometry` GUI
- Change either of the two `Instrument` combo boxes and the other should change in sync

Fixes #23041 

*This does not require release notes* because it only concerns under-the-hood changes. There is no change in behaviour compared to the existing reflectometry GUI.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

TTitcombe commented on qt/scientific_interfaces/ISISReflectometry/GUI/Instrument/InstrumentPresenter.cpp:
```diff
@@ -23,16 +24,33 @@ InstrumentPresenter::InstrumentPresenter(IInstrumentView *view,
   m_view->subscribe(this);
+}
+
+void InstrumentPresenter::acceptMainPresenter(
```
There are a few small chunks of repeated code across the presenters. Would it make sense to move these to a bare-bones class for the various presenters to inherit from? (This may well break some best practice I'm not aware of)

TTitcombe commented on qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsPresenter.cpp:
```diff
@@ -143,14 +134,89 @@ void RunsPresenter::notifyTransfer() {
+    // existing table first
+    // We'll prompt the user to check it's ok to delete existing rows
+
+    // tablePresenter()->setPromptUser(false);
```
I assume this is covered in a later PR?

TTitcombe commented on qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsPresenter.cpp:
```diff
@@ -143,14 +134,89 @@ void RunsPresenter::notifyTransfer() {
+    // existing table first
+    // We'll prompt the user to check it's ok to delete existing rows
+
+    // tablePresenter()->setPromptUser(false);
```
I assume this is covered in a later PR?

gemmaguest replied:
Yes, that's right

TTitcombe commented on qt/scientific_interfaces/ISISReflectometry/GUI/Instrument/InstrumentPresenter.cpp:
```diff
@@ -23,16 +24,33 @@ InstrumentPresenter::InstrumentPresenter(IInstrumentView *view,
   m_view->subscribe(this);
+}
+
+void InstrumentPresenter::acceptMainPresenter(
```
There are a few small chunks of repeated code across the presenters. Would it make sense to move these to a bare-bones class for the various presenters to inherit from? 
p.s. This may well break some best practice I'm not aware of
p.p.s. The total number of repeated lines are small, however, so creating a new class may result in more code

gemmaguest replied:
I'd rather leave this for now, but I think it might be worth doing in a future PR. I might also be able to tidy it up so there's a single interface for all of the tab presenters, instead of a separate interface for each one. I've added a note to #23051 to look into this.

TTitcombe commented on qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsPresenter.cpp:
```diff
@@ -143,14 +134,89 @@ void RunsPresenter::notifyTransfer() {
+}
+
+void RunsPresenter::instrumentChanged(std::string const &instrumentName) {
+  UNUSED_ARG(instrumentName)
```
does the `UNUSED_ARG` flag need to be there since we're passing instrumentName into `setSearchIntrument` and `instrumentChanged`?

gemmaguest replied:
No - I'll take it out :+1:

TTitcombe commented on qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsPresenter.cpp:
```diff
@@ -424,35 +454,17 @@ void RunsPresenter::transfer(const std::set<int> &rowsToTransfer,
-  m_view->setTransferButtonEnabled(/*!isProcessing() &&*/ !isAutoreducing());
-  m_view->setInstrumentComboEnabled(/*!isProcessing() &&*/ !isAutoreducing());
-  m_view->setAutoreducePauseButtonEnabled(isAutoreducing());
+  m_view->setInstrumentComboEnabled(!isProcessing() && !isAutoreducing());
```
`const bool shouldEnable = !isProcessing() && !isAutoreducing();` and then pass `shouldEnable` into the methods to avoid repeated calls to `isProcessing` and `isAutoreducing`?

gemmaguest replied:
I'd rather avoid a vague name like `shouldEnable` because different widgets have different conditions under which they should be enabled. I could call it `notProcessingOrAutoreducing` but that probably wouldn't help with readability!

TTitcombe commented on qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/IRunsTableView.h:
```diff
@@ -38,19 +39,41 @@ be handled by a subclass.
+  virtual std::string getInstrumentName() const = 0;
+  virtual void setInstrumentName(std::string const &instrumentName) = 0;
+
+  virtual void setJobsTableEnabled(bool enable) = 0;
```
can these bools be const'd?

gemmaguest replied:
The local style is not to

gemmaguest commented:
@TTitcombe Thanks for the feedback - I think I've addressed all of the points. I've kept the same Pause button behaviour as the current GUI. It's intended to help indicate the difference between when an autoreduction is running (in which case processing is also running, hence both are enabled) or just that processing alone is running.
